editoast/src/models/projects.rs

```diff
@@ -54,44 +106,91 @@ impl Project {
         Ok(studies_count)
     }
 
+    /// Updates a project's image and deletes the old one if it is not used by another project
+    #[tracing::instrument(skip(conn), ret, err)]
     pub async fn update_and_prune_document(
+        &mut self,
         conn: &mut DbConnection,
-        project_changeset: Changeset<Self>,
-        project_id: i64,
-    ) -> Result<Project> {
-        let old_project =
-            Project::retrieve_or_fail(conn, project_id, || ProjectError::NotFound { project_id })
-                .await?;
-        let image_to_delete = if Some(old_project.image) != project_changeset.image {
-            old_project.image
-        } else {
-            None
-        };
-        let project = project_changeset
-            .update_or_fail(conn, project_id, || ProjectError::NotFound { project_id })
-            .await?;
-        if let Some(image) = image_to_delete {
-            // We don't check the result. We don't want to throw an error if the image is used in another project.
-            let _ = Document::delete_static(conn, image).await;
-        }
-        Ok(project)
+        new_doc_id: Option<i64>,
+    ) -> Result<(), Error> {
+        conn.transaction(|mut conn| {
+            async move {
+                let old_doc_id = self.image;
+                self.image = new_doc_id;
+                self.save(&mut conn).await?;
+                if let Some(doc_id) = old_doc_id {
+                    if new_doc_id.is_none() || new_doc_id.is_some_and(|new| new != doc_id) {
+                        try_delete_document(&conn, doc_id).await?;
+                    }
+                }
+                Ok::<_, Error>(())
+            }
+            .scope_boxed()
+        })
+        .await?;
+        Ok(())
     }
 
-    pub async fn delete_and_prune_document(
-        conn: &mut DbConnection,
+    /// Deletes a project and prunes the image if it is not used by another project
+    #[tracing::instrument(skip(conn), ret, err)]
+    pub async fn delete_and_prune_document(self, conn: &mut DbConnection) -> Result<(), Error> {
+        conn.transaction(|mut conn| {
+            async move {
+                if !self.delete(&mut conn).await? {
+                    tracing::error!(
```

Comment:
nitpick: not really a problem, is it?
```suggestion
                    tracing::warn!(
```

Reply:
agreed 44e570246